// zephyr-agent@0.1.2/dist/lib/transformers/ze-create-manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertResolvedDependencies = convertResolvedDependencies;
exports.createZephyrManifest = createZephyrManifest;
exports.createManifestContent = createManifestContent;
exports.createManifestAsset = createManifestAsset;
const node_crypto_1 = require("node:crypto");
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const logging_1 = require("../logging");
const env_variables_1 = require("../env-variables");
function convertResolvedDependencies(dependencies) {
    return Object.fromEntries(dependencies.map((dep) => [
        dep.name,
        {
            name: dep.name,
            application_uid: dep.application_uid,
            remote_entry_url: dep.remote_entry_url,
            default_url: dep.default_url,
            library_type: dep.library_type,
        },
    ]));
}
/**
 * Creates a zephyr-manifest.json file with resolved dependencies and environment
 * variables
 *
 * @param dependencies - The resolved dependencies from the build
 * @param includeEnvVars - Whether to include environment variables in the manifest
 * @returns The manifest content and the asset object
 */
function createZephyrManifest(dependencies, includeEnvVars = true) {
    const content = createManifestContent(dependencies, includeEnvVars);
    const asset = createManifestAsset(content);
    return { content, asset };
}
function createManifestContent(dependencies, includeEnvVars = true) {
    logging_1.ze_log.manifest('Creating manifest with dependencies:', (dependencies === null || dependencies === void 0 ? void 0 : dependencies.length) || 0);
    // Build the dependencies object
    const dependenciesMap = {};
    dependencies.forEach((dep) => {
        dependenciesMap[dep.name] = {
            name: dep.name,
            application_uid: dep.application_uid,
            remote_entry_url: dep.remote_entry_url,
            default_url: dep.default_url,
            library_type: dep.library_type,
        };
    });
    // Create the manifest object
    const manifest = {
        version: zephyr_edge_contract_1.ZEPHYR_MANIFEST_VERSION,
        timestamp: new Date().toISOString(),
        dependencies: dependenciesMap,
        zeVars: {},
    };
    // Add environment variables if requested
    if (includeEnvVars) {
        const envVars = (0, env_variables_1.collectZEPublicVars)(process.env);
        if (Object.keys(envVars).length > 0) {
            manifest.zeVars = envVars;
            logging_1.ze_log.manifest(`Environment variables: ${Object.keys(envVars).length} ZE_PUBLIC_* vars`);
        }
    }
    logging_1.ze_log.manifest(`Dependencies: ${Object.keys(dependenciesMap).join(', ') || 'none'}`);
    // Convert to JSON string with pretty printing
    return JSON.stringify(manifest, null, 2);
}
function createManifestAsset(content) {
    const contentBuffer = Buffer.from(content);
    // Calculate hash for the content
    const hash = (0, node_crypto_1.createHash)('sha256')
        .update(content + 'zephyr-manifest.json')
        .digest('hex');
    // Return the asset object
    return {
        path: 'zephyr-manifest.json',
        extname: '.json',
        hash,
        size: contentBuffer.length,
        buffer: contentBuffer,
    };
}
