// zephyr-xpack-internal@0.1.2/dist/federation-dashboard-legacy/get-build-stats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuildStats = getBuildStats;
const zephyr_agent_1 = require("zephyr-agent");
const extract_federation_config_1 = require("../xpack-extract/extract-federation-config");
const FederationDashboardPlugin_1 = require("./utils/federation-dashboard-plugin/FederationDashboardPlugin");
async function getBuildStats({ stats, stats_json, pluginOptions, EDGE_URL, DOMAIN, PLATFORM, TYPE, DELIMITER: delimiter = undefined, }) {
    zephyr_agent_1.ze_log.app('get build stats started. create federation dashboard plugin');
    const ze_engine = pluginOptions.zephyr_engine;
    const app = ze_engine.applicationProperties;
    const { git } = ze_engine.gitProperties;
    const { isCI } = ze_engine.env;
    const dashboardPlugin = new FederationDashboardPlugin_1.FederationDashboardPlugin({
        app,
        git,
        context: {
            isCI,
        },
    });
    zephyr_agent_1.ze_log.app('process webpack graph.pluginOptions', pluginOptions);
    const convertedGraph = dashboardPlugin.processWebpackGraph({
        stats,
        stats_json,
        pluginOptions,
    });
    if (!convertedGraph) {
        throw new zephyr_agent_1.ZephyrError(zephyr_agent_1.ZeErrors.ERR_CONVERT_GRAPH_TO_DASHBOARD);
    }
    const version = await ze_engine.snapshotId;
    const application_uid = ze_engine.application_uid;
    const buildId = await ze_engine.build_id;
    const build_target = ze_engine.env.target ?? 'web';
    // todo: add support for multiple federation configs
    const mfConfig = Array.isArray(pluginOptions.mfConfig)
        ? pluginOptions.mfConfig[0]
        : pluginOptions.mfConfig;
    const { name, filename } = mfConfig ? ((0, extract_federation_config_1.extractFederatedConfig)(mfConfig) ?? {}) : {};
    const remotes = ze_engine.federated_dependencies;
    const data_overrides = {
        id: application_uid,
        name: name,
        edge: { url: EDGE_URL, delimiter },
        domain: DOMAIN,
        platform: PLATFORM,
        type: TYPE,
        app: Object.assign({}, app, { buildId }),
        version,
        git,
        remote: filename,
        remotes: remotes?.map(({ application_uid }) => application_uid) ?? [],
        context: { isCI },
        build_target,
        zephyrDependencies: ze_engine.zephyr_dependencies,
    };
    // todo: extend data
    const res = Object.assign({}, convertedGraph, data_overrides, {
        project: name, // Add missing project property
        tags: [], // Add missing tags property with empty array as default
    });
    zephyr_agent_1.ze_log.app('get build stats done.', res);
    return res;
}
