// zephyr-agent@0.1.2/dist/lib/logging/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ze_log = exports.ze_debug = exports.ze_error = exports.brightRedBgName = exports.brightGreenBgName = exports.brightYellowBgName = exports.brightBlueBgName = exports.dimmedName = void 0;
// light weight functions for decorated console.error + able to toggle different parts of whole module's logging result
const debug_1 = require("debug");
const picocolor_1 = require("./picocolor");
const file_logger_1 = require("./file-logger");
//TODO: this should be traced and logged into new relic
const name = ' ZEPHYR ';
exports.dimmedName = (0, picocolor_1.dim)(name);
exports.brightBlueBgName = (0, picocolor_1.bold)((0, picocolor_1.bgCyanBright)((0, picocolor_1.black)(name)));
exports.brightYellowBgName = (0, picocolor_1.bold)((0, picocolor_1.bgYellowBright)((0, picocolor_1.black)(name)));
exports.brightGreenBgName = (0, picocolor_1.bold)((0, picocolor_1.bgGreenBright)((0, picocolor_1.black)(name)));
exports.brightRedBgName = (0, picocolor_1.bold)((0, picocolor_1.bgRedBright)((0, picocolor_1.black)(name)));
function wrapDebugLogger(logger, context) {
    const wrappedFn = (...args) => {
        const message = args
            .map((arg) => (typeof arg === 'object' ? JSON.stringify(arg) : String(arg)))
            .join(' ');
        // Always write to file if file logging is enabled, regardless of DEBUG env var
        if ((0, file_logger_1.isFileLoggingEnabled)()) {
            (0, file_logger_1.writeLogToFile)({
                level: 'debug',
                message,
                action: `debug:${context}`,
                timestamp: Date.now(),
            });
        }
        // Only output to console if DEBUG env var matches
        if (!logger.enabled)
            return;
        logger(...args);
    };
    return Object.assign(wrappedFn, {
        enabled: logger.enabled,
        namespace: logger.namespace,
    });
}
const rawLoggers = {
    app: (0, debug_1.debug)('zephyr:app'),
    auth: (0, debug_1.debug)('zephyr:auth'),
    buildstats: (0, debug_1.debug)('zephyr:buildstats'),
    config: (0, debug_1.debug)('zephyr:config'),
    git: (0, debug_1.debug)('zephyr:git'),
    http: (0, debug_1.debug)('zephyr:http'),
    init: (0, debug_1.debug)('zephyr:init'),
    manifest: (0, debug_1.debug)('zephyr:manifest'),
    mf: (0, debug_1.debug)('zephyr:mf'),
    misc: (0, debug_1.debug)('zephyr:misc'),
    package: (0, debug_1.debug)('zephyr:package'),
    remotes: (0, debug_1.debug)('zephyr:remotes'),
    snapshot: (0, debug_1.debug)('zephyr:snapshot'),
    upload: (0, debug_1.debug)('zephyr:upload'),
    debug: (0, debug_1.debug)('zephyr:debug'),
    error: (0, debug_1.debug)('zephyr:error'),
};
exports.ze_error = wrapDebugLogger(rawLoggers.error, 'error');
exports.ze_debug = wrapDebugLogger(rawLoggers.debug, 'debug');
const createLogger = () => {
    return {
        app: wrapDebugLogger(rawLoggers.app, 'app'),
        auth: wrapDebugLogger(rawLoggers.auth, 'auth'),
        buildstats: wrapDebugLogger(rawLoggers.buildstats, 'buildstats'),
        config: wrapDebugLogger(rawLoggers.config, 'config'),
        git: wrapDebugLogger(rawLoggers.git, 'git'),
        http: wrapDebugLogger(rawLoggers.http, 'http'),
        init: wrapDebugLogger(rawLoggers.init, 'init'),
        manifest: wrapDebugLogger(rawLoggers.manifest, 'manifest'),
        mf: wrapDebugLogger(rawLoggers.mf, 'mf'),
        misc: wrapDebugLogger(rawLoggers.misc, 'misc'),
        package: wrapDebugLogger(rawLoggers.package, 'package'),
        remotes: wrapDebugLogger(rawLoggers.remotes, 'remotes'),
        snapshot: wrapDebugLogger(rawLoggers.snapshot, 'snapshot'),
        upload: wrapDebugLogger(rawLoggers.upload, 'upload'),
        debug: wrapDebugLogger(rawLoggers.debug, 'debug'),
        error: wrapDebugLogger(rawLoggers.error, 'error'),
    };
};
/**
 * Debug contexts:
 *
 * - Ze_log.app: Application config information
 * - Ze_log.auth: Authentication and token management
 * - Ze_log.buildstats: Build information for Dashboard API
 * - Ze_log.config: Authentication and token management
 * - Ze_log.git: Git configuration and provider
 * - Ze_log.http: http requests
 * - Ze_log.manifest: Generation and inclusion of zephyr manifest
 * - Ze_log.init: Initialization and setup operations
 * - Ze_log.manifest: Manifest generation and processing
 * - Ze_log.mf: Module Federation config
 * - Ze_log.misc: Miscellaneous
 * - Ze_log.package: Package.json parsing
 * - Ze_log.remotes: Remote dependency resolution
 * - Ze_log.snapshot: Snapshot publish
 * - Ze_log.upload: Asset and build stats upload
 */
exports.ze_log = createLogger();
// If debug mode is not enabled just print whatever console output is
// If debug mode is enabled print the error from our end
