// zephyr-agent@0.1.2/dist/lib/deployment/aws-upload.strategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.awsUploadStrategy = awsUploadStrategy;
exports.createBucket = createBucket;
const logging_1 = require("../logging");
const errors_1 = require("../errors");
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const http_request_1 = require("../http/http-request");
const edge_actions_1 = require("../edge-actions");
const upload_base_1 = require("./upload-base");
const EDGE_LAMBDA_MAX_BODY_SIZE = 788403; // floor(1024 * 1024 / 1.33), where 1.33 is size overhead of base64 encoding
async function awsUploadStrategy(zephyr_engine, { snapshot, getDashData, assets: { assetsMap, missingAssets } }) {
    const snapshotSize = Buffer.byteLength(JSON.stringify(snapshot), 'utf8');
    if (snapshotSize > EDGE_LAMBDA_MAX_BODY_SIZE) {
        const opts = {
            entity_name: 'snapshot',
            entity_size: snapshotSize,
            max_allowed_size: EDGE_LAMBDA_MAX_BODY_SIZE,
        };
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_MAX_PAYLOAD_SIZE_EXCEEDED, opts);
    }
    for (const missingAsset of missingAssets) {
        if (missingAsset.size > EDGE_LAMBDA_MAX_BODY_SIZE) {
            const opts = {
                entity_name: missingAsset.path,
                entity_size: missingAsset.size,
                max_allowed_size: EDGE_LAMBDA_MAX_BODY_SIZE,
            };
            throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_MAX_PAYLOAD_SIZE_EXCEEDED, opts);
        }
    }
    await createBucket(zephyr_engine.application_uid);
    await (0, upload_base_1.uploadAssets)(zephyr_engine, { assetsMap, missingAssets });
    const versionUrl = await (0, edge_actions_1.zeUploadSnapshot)(zephyr_engine, { snapshot });
    // Waits for the reply to check upload problems, but the reply is a simply
    // 200 OK sent before any processing
    await (0, upload_base_1.uploadBuildStatsAndEnableEnvs)(zephyr_engine, { getDashData, versionUrl });
    return versionUrl;
}
async function createBucket(application_uid) {
    const { EDGE_URL, jwt } = await (0, get_application_configuration_1.getApplicationConfiguration)({
        application_uid,
    });
    const options = {
        method: 'POST',
        headers: {
            can_write_jwt: jwt,
        },
    };
    const url = new URL('/upload', EDGE_URL);
    url.searchParams.append('type', 'bucket');
    const [ok, cause] = await (0, http_request_1.makeRequest)(url, options);
    if (!ok) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_FAILED_UPLOAD, {
            type: 'bucket',
            cause,
        });
    }
    logging_1.ze_log.snapshot('Done: bucket initialized...');
}
