// zephyr-agent@0.1.2/dist/lib/env-variables/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectZEPublicVars = exports.calculateManifestHash = exports.generateManifestContent = exports.buildEnvJsonAsset = exports.injectBeforeHeadClose = exports.buildModulePreload = exports.injectImportMap = exports.buildImportMap = exports.rewriteEnvReadsToVirtualModule = exports.detectEnvReads = exports.VIRTUAL_SPECIFIER = void 0;
exports.buildEnvImportMap = buildEnvImportMap;
exports.buildEnvModuleSource = buildEnvModuleSource;
exports.buildEnvImportMapScript = buildEnvImportMapScript;
var env_var_rewrites_1 = require("./env-var-rewrites");
Object.defineProperty(exports, "VIRTUAL_SPECIFIER", { enumerable: true, get: function () { return env_var_rewrites_1.VIRTUAL_SPECIFIER; } });
Object.defineProperty(exports, "detectEnvReads", { enumerable: true, get: function () { return env_var_rewrites_1.detectEnvReads; } });
Object.defineProperty(exports, "rewriteEnvReadsToVirtualModule", { enumerable: true, get: function () { return env_var_rewrites_1.rewriteEnvReadsToVirtualModule; } });
Object.defineProperty(exports, "buildImportMap", { enumerable: true, get: function () { return env_var_rewrites_1.buildImportMap; } });
Object.defineProperty(exports, "injectImportMap", { enumerable: true, get: function () { return env_var_rewrites_1.injectImportMap; } });
Object.defineProperty(exports, "buildModulePreload", { enumerable: true, get: function () { return env_var_rewrites_1.buildModulePreload; } });
Object.defineProperty(exports, "injectBeforeHeadClose", { enumerable: true, get: function () { return env_var_rewrites_1.injectBeforeHeadClose; } });
Object.defineProperty(exports, "buildEnvJsonAsset", { enumerable: true, get: function () { return env_var_rewrites_1.buildEnvJsonAsset; } });
Object.defineProperty(exports, "generateManifestContent", { enumerable: true, get: function () { return env_var_rewrites_1.generateManifestContent; } });
Object.defineProperty(exports, "calculateManifestHash", { enumerable: true, get: function () { return env_var_rewrites_1.calculateManifestHash; } });
Object.defineProperty(exports, "collectZEPublicVars", { enumerable: true, get: function () { return env_var_rewrites_1.collectZEPublicVars; } });
function buildEnvImportMap(appUid, remotes) {
    const imports = {};
    // Add the main env module mapping - now points to zephyr-manifest.json
    imports[`env:vars:${appUid}`] = '/zephyr-manifest.json';
    // Add environment variable manifest entries for remotes
    // Note: Module Federation remotes are loaded by MF runtime, not through import maps
    remotes.forEach((remote) => {
        if (remote.application_uid && remote.remote_entry_url) {
            // Environment variables manifest for the remote
            try {
                const urlStr = remote.remote_entry_url.includes('@')
                    ? remote.remote_entry_url.split('@')[1]
                    : remote.remote_entry_url;
                const origin = new URL(urlStr).origin;
                imports[`env:vars:${remote.application_uid}`] = `${origin}/zephyr-manifest.json`;
            }
            catch (_a) {
                // If URL parsing fails, skip the env:vars entry
                console.warn(`Failed to parse remote URL for env vars: ${remote.remote_entry_url}`);
            }
        }
    });
    return imports;
}
function buildEnvModuleSource(applicationUid) {
    // Collect all ZE_PUBLIC_* environment variables
    const envVars = {};
    for (const [key, value] of Object.entries(process.env)) {
        if (key.startsWith('ZE_PUBLIC_') && typeof value === 'string') {
            envVars[key] = value;
        }
    }
    // Generate the module source code
    const exports = Object.entries(envVars)
        .map(([key, value]) => `export const ${key} = ${JSON.stringify(value)};`)
        .join('\n');
    // Include default export with all vars and the app UID
    const defaultExport = `
export default {
  APPLICATION_UID: ${JSON.stringify(applicationUid)},
  ${Object.entries(envVars)
        .map(([key, value]) => `${key}: ${JSON.stringify(value)}`)
        .join(',\n  ')}
};`;
    return `// Zephyr Environment Variables Module
// Auto-generated for app: ${applicationUid}
${exports}
${defaultExport}
`;
}
function buildEnvImportMapScript(appUid, remotes) {
    const importMap = {
        imports: buildEnvImportMap(appUid, remotes),
    };
    return `<script type="importmap">${JSON.stringify(importMap, null, 2)}</script>`;
}
