// zephyr-agent@0.1.2/dist/lib/node-persist/app-deploy-result-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setAppDeployResult = setAppDeployResult;
exports.getAppDeployResult = getAppDeployResult;
exports.removeAppDeployResult = removeAppDeployResult;
exports.getAllDeployedApps = getAllDeployedApps;
exports.getAllAppDeployResults = getAllAppDeployResults;
const node_persist_1 = require("node-persist");
const storage_1 = require("./storage");
const storage_keys_1 = require("./storage-keys");
async function setAppDeployResult(application_uid, value) {
    await storage_1.storage;
    void (await (0, node_persist_1.setItem)(`${storage_keys_1.StorageKeys.ze_app_deploy_result}:${application_uid}`, value, {
        ttl: 1000 * 60 * 60 * 24,
    }));
}
async function getAppDeployResult(application_uid) {
    await storage_1.storage;
    return (0, node_persist_1.getItem)(`${storage_keys_1.StorageKeys.ze_app_deploy_result}:${application_uid}`);
}
async function removeAppDeployResult(application_uid) {
    await storage_1.storage;
    await (0, node_persist_1.removeItem)(`${storage_keys_1.StorageKeys.ze_app_deploy_result}:${application_uid}`);
}
async function getAllDeployedApps() {
    await storage_1.storage;
    const allKeys = await (0, node_persist_1.keys)();
    const resultKeys = allKeys.filter((key) => key.startsWith(storage_keys_1.StorageKeys.ze_app_deploy_result));
    return resultKeys.map((key) => key.substring(storage_keys_1.StorageKeys.ze_app_deploy_result.length + 1));
}
async function getAllAppDeployResults() {
    await storage_1.storage;
    const results = {};
    await (0, node_persist_1.forEach)((entry) => {
        if (entry.key && entry.key.startsWith(storage_keys_1.StorageKeys.ze_app_deploy_result)) {
            const application_uid = entry.key.substring(storage_keys_1.StorageKeys.ze_app_deploy_result.length + 1);
            results[application_uid] = entry.value;
        }
    });
    return results;
}
