// zephyr-agent@0.1.2/dist/lib/node-persist/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveToken = saveToken;
exports.getToken = getToken;
exports.removeToken = removeToken;
exports.cleanTokens = cleanTokens;
const node_persist_1 = require("node-persist");
const secret_token_1 = require("./secret-token");
const storage_1 = require("./storage");
const storage_keys_1 = require("./storage-keys");
const http_request_1 = require("../http/http-request");
const server_token_1 = require("./server-token");
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const user_email_1 = require("./user-email");
const debug_1 = require("../logging/debug");
async function saveToken(token) {
    await storage_1.storage;
    await (0, node_persist_1.setItem)(storage_keys_1.StorageKeys.ze_auth_token, token);
}
async function getToken(git_config) {
    const tokenFromEnv = (0, secret_token_1.getSecretToken)();
    const server_token = (0, server_token_1.getServerToken)();
    if (tokenFromEnv) {
        return tokenFromEnv;
    }
    await storage_1.storage;
    const token = await (0, node_persist_1.getItem)(storage_keys_1.StorageKeys.ze_auth_token);
    if (token) {
        return token;
    }
    if (server_token) {
        if (!git_config) {
            debug_1.ze_log.error('No git config provided, skipping server token check');
            return undefined;
        }
        return await getTokenFromServerToken(server_token, git_config.git.email);
    }
    return undefined;
}
async function removeToken() {
    await storage_1.storage;
    await (0, node_persist_1.removeItem)(storage_keys_1.StorageKeys.ze_auth_token);
}
async function cleanTokens() {
    await storage_1.storage;
    await (0, node_persist_1.clear)();
}
async function getTokenFromServerToken(server_token, git_email) {
    var _a, _b;
    const email = (_a = (0, user_email_1.getUserEmail)()) !== null && _a !== void 0 ? _a : git_email;
    const [ok, cause, data] = await (0, http_request_1.makeRequest)({
        path: zephyr_edge_contract_1.ze_api_gateway.get_access_token_by_server_token,
        base: (0, zephyr_edge_contract_1.ZE_API_ENDPOINT)(),
        query: { email },
    }, {
        headers: {
            Authorization: `Bearer ${server_token}`,
        },
    });
    if (!ok) {
        if (cause instanceof Error) {
            debug_1.ze_log.error('Failed to get token from server token:', cause.message);
        }
        else {
            debug_1.ze_log.error('Failed to get token from server token:', cause);
        }
        return undefined;
    }
    await saveToken((_b = data === null || data === void 0 ? void 0 : data.access_token) !== null && _b !== void 0 ? _b : '');
    return data === null || data === void 0 ? void 0 : data.access_token;
}
