// zephyr-agent@0.1.2/dist/lib/http/upload-snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadSnapshot = uploadSnapshot;
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const errors_1 = require("../errors");
const logging_1 = require("../logging");
const http_request_1 = require("./http-request");
async function uploadSnapshot({ body, application_uid, }) {
    const { EDGE_URL, jwt, ENVIRONMENTS } = await (0, get_application_configuration_1.getApplicationConfiguration)({
        application_uid,
    });
    const json = JSON.stringify(body);
    logging_1.ze_log.snapshot('Sending snapshot to edge:', JSON.stringify(body, null, 2));
    const resp = await doUploadSnapshotRequest({ json, edge_url: EDGE_URL, jwt });
    if (ENVIRONMENTS != null) {
        const env_edge_urls = Array.from(new Set(Object.values(ENVIRONMENTS).filter((envCfg) => envCfg.edgeUrl !== EDGE_URL)));
        await Promise.all(env_edge_urls.map((envConfig) => {
            return doUploadSnapshotRequest({ json, edge_url: envConfig.edgeUrl, jwt });
        }));
    }
    logging_1.ze_log.snapshot('Done: snapshot uploaded...', body);
    return resp;
}
async function doUploadSnapshotRequest({ json, edge_url, jwt, }) {
    const options = {
        method: 'POST',
        headers: {
            'Content-Length': Buffer.byteLength(json).toString(),
            'Content-Type': 'application/json; charset=utf-8',
            can_write_jwt: jwt,
        },
    };
    const url = new URL('/upload', edge_url);
    url.searchParams.append('type', 'snapshot');
    url.searchParams.append('skip_assets', 'true');
    logging_1.ze_log.snapshot('Upload URL:', url.toString());
    const [ok, cause, resp] = await (0, http_request_1.makeRequest)(url, options, json);
    if (!ok) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_FAILED_UPLOAD, {
            type: 'snapshot',
            cause,
        });
    }
    return resp;
}
